### PR TITLE
[new-formatter] TAP formatter

### PR DESCRIPTION
**Pull request by LKay**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:

Adds new TAP formatter for failures output. Can be used with [any TAP reporter](https://github.com/sindresorhus/awesome-tap#reporters)

#### CHANGELOG.md entry:

- [enhancement] New TAP formatter

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @LKay! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/formatters/tapFormatter.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright 2013 Palantir Technologies, Inc.
```
Comment (ajafff) [inline src/formatters/tapFormatter.ts]: s/2013/2017/

File: src/formatters/tapFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                `  severity: ${failureSeverity}`,
+                "  data:",
+                `    ruleName: ${ruleName}`,
+                `    fileName: ${fileName}`,
+                `    line: ${lineAndCharacter.line}`,
+                `    character: ${lineAndCharacter.character}`,
+                `    failureString: ${failureString}`,
+                `    rawLines: ${failureRaw}`,
+                "  ...",
+            ].join("\n");
```
Comment (ajafff) [inline src/formatters/tapFormatter.ts]: Just a suggestion: use multiline template strings and `Utils.dedent`

```ts
            return Utils.dedent`
                not ok ${i + 1} - ${failureMessage}
                  ---
                  message : ${failureMessage}
                  severity: ${failureSeverity}
                  data:
                    ruleName: ${ruleName}
                    fileName: ${fileName}
                    line: ${lineAndCharacter.line}
                    character: ${lineAndCharacter.character}
                    failureString: ${failureString}
                    rawLines: ${failureRaw}
                  ...`;
```

**Comment by nchen63** [comment]
@LKay thanks!

**Pull request merged**